dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/handling/DataHandler.java

```diff
@@ -0,0 +1,1161 @@
+        List<Indicator> indicators )
+    {
+        List<DimensionalItemObject> items = newArrayList(
+            expressionService.getIndicatorDimensionalItemObjects( resolveIndicatorExpressions( indicators ) ) );
+
+        if ( items.isEmpty() )
+        {
+            return new ArrayListValuedHashMap<>();
+        }
+
+        items = replaceOperandTotalsWithDataElements( items );
+
+        DimensionalObject dimension = new BaseDimensionalObject( DATA_X_DIM_ID, DATA_X, null, DISPLAY_NAME_DATA_X,
+            items );
+
+        DataQueryParams dataSourceParams = newBuilder( params )
+            .replaceDimension( dimension )
+            .withMeasureCriteria( new HashMap<>() )
+            .withIncludeNumDen( false )
+            .withSkipHeaders( true )
+            .withOutputFormat( ANALYTICS )
+            .withSkipMeta( true ).build();
+
+        // Each row in the Grid contains: dimension uid | period | value
+        Grid grid = dataAggregator.getAggregatedDataValueGridInternal( dataSourceParams );
+        MultiValuedMap<String, DimensionItemObjectValue> result = new ArrayListValuedHashMap<>();
+
+        if ( isEmpty( grid.getRows() ) )
+        {
+            return result;
+        }
+
+        BiFunction<Integer, Integer, Integer> replaceIndexIfMissing = ( Integer index,
+            Integer defaultIndex ) -> index == -1 ? defaultIndex : index;
+
+        final int dataIndex = replaceIndexIfMissing.apply( grid.getIndexOfHeader( DATA_X_DIM_ID ), 0 );
+        final int periodIndex = replaceIndexIfMissing.apply( grid.getIndexOfHeader( PERIOD_DIM_ID ), 1 );
+        final int valueIndex = grid.getWidth() - 1;
+
+        final List<DimensionalItemObject> basePeriods = params.getPeriods();
+
+        for ( List<Object> row : grid.getRows() )
+        {
+            final List<DimensionalItemObject> dimensionalItems = findDimensionalItems( (String) row.get( dataIndex ),
+                items );
+
+            // Check if the current row's Period belongs to the list of periods from the
+            // original Analytics request.
+            // The row may not have a Period if Period is used as filter.
+            if ( hasPeriod( row, periodIndex ) && isPeriodInPeriods( (String) row.get( periodIndex ), basePeriods ) )
+            {
+                if ( dimensionalItems.size() == 1 )
+                {
+                    addItemBasedOnPeriodOffset( grid, result, periodIndex, valueIndex, row, dimensionalItems );
+                }
+            }
+            else
+            {
+                if ( dimensionalItems.size() == 1 )
+                {
+                    result.put( dimensionalItems.get( 0 ).getDimensionItem(),
+                        new DimensionItemObjectValue( dimensionalItems.get( 0 ), (Double) row.get( valueIndex ) ) );
+                }
+            }
+        }
+
+        return result;
+    }
+
+    /**
+     * Add the given Indicator values to the Grid.
+     * 
+     * @param params
```

Comment:
Add missing javadoc for the method arguments or remove `@param`

Reply:
Done